[W.I.P] Documentation for Viper by Example

NedYork opened the pull request:
### - What I did

In this PR, I wrote up the initial drafts for "Viper by Example" for 3 examples so far - 1) Voting, 2) Simple Open Auction, and 3) Safe Remote Purchase. If any information is misstated or ambiguous, please let me know.

### - How I did it

I created a new file called viper-by-example.rst in the docs folder and wrote it restructured text format.

### - How to verify it

N/A

### - Description for the changelog

New documentation for 'Viper by Example'.

### - Cute Animal Picture

![image](https://user-images.githubusercontent.com/5008104/31833862-e6c32412-b599-11e7-9ced-a8455145e914.png)

I would laugh so hard if someone wore these at Devcon3 haha.

chgue commented:
Just skimmed it, but it looks pretty good! :) 
Two small suggestions: 
- Firstly, I think ordering the contracts by complexity (In my opinion: auction, safe remote purchase, and then voting) would be beneficial. 
- Secondly, you could try using [literalinclude](http://www.sphinx-doc.org/en/1.5.1/markup/code.html#includes) to display the code. However, I have never tried it, but I think it could be pretty useful since it removes the need to update the docs if e.g. a small bug is fixed in an example.

DavidKnott commented:
To make this included in the docs when `make docs` is run please add the `viper-by-example.rst` to the index.

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,633 @@
+transaction between a buyer and a seller. In this system, a seller posts an item
+for sale and makes a deposit to the contract of twice the item's ``value``. At
+this moment, the contract has a balance of 2 * ``value``. The seller can reclaim
+the deposit and close the sale as long as a buyer had not yet made a purchase.
```
Change `had` to `has`

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,633 @@
+for sale and makes a deposit to the contract of twice the item's ``value``. At
+this moment, the contract has a balance of 2 * ``value``. The seller can reclaim
+the deposit and close the sale as long as a buyer had not yet made a purchase.
+If a buyer is interested in making a purchase, he/she would make a payment and
```
Change `he/she` to `they` to be safe.

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,633 @@
+      selfdestruct(self.seller) #Refunds seller, deletes contract
+
+The ``abort()`` method is a method only callable by the seller and while the
+contract is still ``unlocked`` - meaning it callable only prior to any buyer
```
`it is callable` (add is)

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,633 @@
+
+The ``abort()`` method is a method only callable by the seller and while the
+contract is still ``unlocked`` - meaning it callable only prior to any buyer
+making a purchase. As we will see in the ``purchase()`` method that as soon as
```
`method, as soon` (take out that)

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,633 @@
+      self.unlocked = False
+
+Like the constructor, the ``purchase()`` method has a ``@payable`` decorator,
+meaning it must be called with a payment. For the buyer to make a valid
```
Same thing as above, `@payable` only allows ether to be sent.

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,633 @@
+      send(self.buyer, self.value) #Return deposit (=value) to buyer
+      selfdestruct(self.seller) #Returns deposit (=2*value) and the purchase price (=value)
+
+Finally, upon the buyer's receipt of the item, the buyer can confirm his/her
```
use `their` instead of `his/her`

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,633 @@
+By calling ``received()``, we begin by ``assert``ing that the contract is indeed
+locked, ensuring that a buyer had previously paid. We also ensure that this
+method is only callable by the buyer. If these two ``assert`` statements pass,
+we refund the buyer his/her initial deposit and send the seller the remain
```
Use `their` instead of `his/her`

DavidKnott replied:
`remaining` instead of `remain`

NedYork commented:
Hey @chgue and @DavidKnott , thanks for the feedback. I've made most of the changes according to both your comments.

To address the ordering of the examples, I'm totally with you. I've structured it as such to parallel the docs of Solidity. I will reorder them soon. As for the literalinclude, there's a few comments in the actual files that I'd like to keep out so for now, I'll leave it as is. Perhaps when all is finalized, I'll do a literalinclude.

Thanks guys!

fubuloubu commented:
2nd on the literalinclude. Please do not duplicate the source files (even if to describe sections of content) as it makes it more difficult to maintain

NedYork commented:
Awesome. I've implemented literalinclude, reworked the ordering, and added two new examples in the past few commits. Will appreciate any feedback! Thanks!

fubuloubu commented:
Can you squash these 19 commits into 1?

DavidKnott commented:
+ 1 For rebasing, it'll also get travis passing

NedYork commented:
Yep. Squashed.

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,628 @@
+  :language: python
+  :lines: 24-38
+
+The ``@payable`` decorator will allow a user to send some ethers to the
```
Change `ethers` to `ether`

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,628 @@
+to their desired bid (not including gas fees). When calling any method within a
+contract, we are provided with a built-in variable ``msg`` and we can access
+the public address of any method caller with ``msg.sender``. Similarly, the
+amount of ether a user sends can be accessed by calling ``msg.sender``.
```
Change `msg.sender` to `msg.value`

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,628 @@
+to see if the new bid is greater than the highest bid. If both ``assert``
+statements pass, we continue to the next lines; otherwise, the ``bid()`` method
+will throw an error and revert the transaction. If the two ``assert`` statements
+the check that the previous bid is not equal to zero pass, we can safely
```
I'd change it to just `If the two ``assert`` statements pass, we can safely`

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,628 @@
+corresponding names. Additionally, a ``self.deadline`` is initialized to set
+a definitive end time for the crowdfunding period.
+
+Now lets take a look at the method of how a person can participate in the
```
Take out `the method of`

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,628 @@
+
+.. note:: Notice that we have access to the total amount sent to the contract by
+calling ``self.balance``, a variable we never explicitly set. Similar to ``msg``
+and ``block``, ``self.balance`` is a variable thats comes free in all Viper
```
Change `free` to `with` as it still costs gas

DavidKnott commented on docs/viper-by-example.rst:
```diff
@@ -0,0 +1,628 @@
+large number of of participants, multiple calls to ``refund()`` may be
+necessary.
+
+One thing to note about the ``send()`` method is that it costs gas. Thus,
```
Refunding costs the person sending the transaction, not the contract itself.

DavidKnott commented:
LGTM
